pkg/compactor/executor.go

```diff
@@ -283,8 +350,149 @@ func (e *schedulerExecutor) updateJobStatus(ctx context.Context, key *compactors
+		blockID, err := ulid.Parse(string(id))
+		if err != nil {
+			level.Error(userLogger).Log("msg", "invalid block ID, abandoning job", "tenant", userID, "err", err)
+			return compactorschedulerpb.ABANDON, errors.Wrapf(err, "failed to parse block ID")
+		}
+		blockIDs = append(blockIDs, blockID)
+	}
+
+	if err := syncer.SyncRequestedMetas(ctx, blockIDs); err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to sync metas")
+	}
+
+	jobMetas, err := getJobMetas(syncer, blockIDs)
+	if err != nil {
+		level.Error(userLogger).Log("msg", "blocks not found after sync, abandoning job", "err", err)
+		return compactorschedulerpb.ABANDON, err
+	}
+
+	var splitNumShards uint32
+	if spec.Job.Split {
+		splitNumShards = uint32(c.cfgProvider.CompactorSplitAndMergeShards(userID))
+	}
+
+	job, err := buildCompactionJobFromMetas(userID, key.Id, jobMetas, spec, splitNumShards)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, err
+	}
+
+	compactor, err := c.newBucketCompactor(ctx, userID, userLogger, userBucket, syncer, reg)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create bucket compactor")
+	}
+
+	// Track that a compaction job has started
+	compactor.metrics.groupCompactionRunsStarted.Inc()
+
+	level.Info(userLogger).Log("msg", "executing compaction job from scheduler", "tenant", userID, "blocks", len(blockIDs), "split", spec.Job.Split)
+	_, compactedBlockIDs, err := compactor.runCompactionJob(ctx, job)
+	if err == nil {
+		compactor.metrics.groupCompactionRunsCompleted.Inc()
+		if hasNonZeroULIDs(compactedBlockIDs) {
+			compactor.metrics.groupCompactions.Inc()
+		}
+		level.Info(userLogger).Log("msg", "compaction job completed", "tenant", userID, "compacted_blocks", len(compactedBlockIDs))
+		return compactorschedulerpb.COMPLETE, nil
+	}
+
+	// At this point the compaction has failed. Track the failure.
+	compactor.metrics.groupCompactionRunsFailed.Inc()
+
+	if ok, issue347Err := isIssue347Error(err); ok {
+		level.Warn(userLogger).Log("msg", "detected issue347 error during compaction", "block", issue347Err.id, "err", err)
+		repairErr := repairIssue347(ctx, userLogger, userBucket, syncer.metrics.blocksMarkedForDeletion, issue347Err)
+		if repairErr == nil {
+			level.Info(userLogger).Log("msg", "successfully repaired issue347 block, job can be re-planned", "block", issue347Err.id)
+			return compactorschedulerpb.COMPLETE, nil
+		}
+		level.Error(userLogger).Log("msg", "failed to repair issue347 block, abandoning job", "block", issue347Err.id, "err", repairErr)
+		return compactorschedulerpb.ABANDON, errors.Wrap(repairErr, "failed to repair issue347 block")
+	}
+
+	if ok, outOfOrderChunksErr := IsOutOfOrderChunkError(err); ok && compactor.skipUnhealthyBlocks {
+		level.Warn(userLogger).Log("msg", "detected out-of-order chunks error during compaction", "block", outOfOrderChunksErr.id, "err", err)
+		markErr := e.retryable.WithContext(ctx).Run(func() error {
```

Comment:
`e.retryable` has a maxRetries of `-1` (infinite). Does that work here, or would we want a different, less aggressive retry policy for these error conditions? It's fine if we have different retry policies for different areas of code, and they don't all need to be configurable.